orbit/pkg/installer/installer.go

```diff
@@ -0,0 +1,233 @@
+	OsqueryClient QueryClient
+	OrbitClient   Client
+
+	// osquerySocketPaht is used to establish the osquery connection
+	// if it's ever lost or disconnected
+	osquerySocketPaht string
+
+	// tempDirFn is the function to call to get the temporary directory to use,
+	// inside of which the script-specific subdirectories will be created. If nil,
+	// the user's temp dir will be used (can be set to t.TempDir in tests).
+	tempDirFn func() string
+
+	// execCmdFn can be set for tests to mock actual execution of the script. If
+	// nil, execCmd will be used, which has a different implementation on Windows
+	// and non-Windows platforms.
+	execCmdFn func(ctx context.Context, scriptPath string, env []string) ([]byte, int, error)
+
+	// can be set for tests to replace os.RemoveAll, which is called to remove
+	// the script's temporary directory after execution.
+	removeAllFn func(string) error
+
+	connectOsquery func(*Runner) error
+}
+
+func NewRunner(client Client, socketPath string, queryTimeout time.Duration) (*Runner, error) {
+	r := &Runner{
+		OrbitClient:       client,
+		osquerySocketPaht: socketPath,
+	}
+
+	return r, nil
+}
+
+func (r *Runner) Run(config *fleet.OrbitConfig) error {
+	if r.connectOsquery == nil {
+		r.connectOsquery = connectOsquery
+	}
+
+	if err := r.connectOsquery(r); err != nil {
+		return fmt.Errorf("software installer runner connecting to osquery: %w", err)
+	}
+	return r.run(context.Background(), config)
+}
+
+func connectOsquery(r *Runner) error {
+	if r.OsqueryClient == nil {
+		osqueryClient, err := osquery.NewClient(r.osquerySocketPaht, 2*time.Second)
+		if err != nil {
+			log.Err(err).Msg("establishing osquery connection for software install runner")
+			return err
+		}
+
+		r.OsqueryClient = osqueryClient.Client
+	}
+
+	return nil
+}
+
+func (r *Runner) run(ctx context.Context, config *fleet.OrbitConfig) error {
+	var errs []error
+	for _, installerID := range config.Notifications.PendingSoftwareInstallerIDs {
+		if ctx.Err() == nil {
+			payload, err := r.installSoftware(ctx, installerID)
+			if err != nil {
+				errs = append(errs, err)
+			}
+			if err := r.OrbitClient.SaveInstallerResult(payload); err != nil {
+				errs = append(errs, ctxerr.Wrap(ctx, err, "saving software install results"))
+			}
+		}
+	}
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
+
+	return nil
+}
+
+func (r *Runner) preConditionCheck(ctx context.Context, query string) (bool, string, error) {
+	res, err := r.OsqueryClient.Query(ctx, query)
+	if err != nil {
+		return false, "", ctxerr.Wrap(ctx, err, "precondition check")
+	}
+
+	if res.Status == nil {
+		return false, "", errors.New("no query status")
+	}
+
+	response, err := json.Marshal(res.Response)
+	if err != nil {
+		return false, "", ctxerr.Wrap(ctx, err, "marshalling query response")
+	}
+
+	if res.Status.Code != 0 {
+		return false, string(response), ctxerr.Wrap(ctx, fmt.Errorf("non-zero query status: %d \"%s\"", res.Status.Code, res.Status.Message))
```

Comment:
do you know what's `res.Response` when `res.Status.Code != 0`? I'm wondering if we should also include `res.Status.Message` in whatever we send to the server or not.

eg: what if you try to run a query for a table that doesn't exist? do you get an error message as part of `Response` or as part of `Status`?

Reply:
Good question, since we don't actually return the error to the server... I suppose I could combine them into a single response and return that

Reply:
You're absolutely right, the error message is in the status message

Reply:
Updated 👍